src/sqlfluff/rules/L007.py

```diff
@@ -58,51 +68,103 @@ def _eval(self, context: RuleContext) -> LintResult:
-                                anchor = context.segment
-                                # TODO: Work out a nice fix for this failure.
-                elif memory["last_code"] and memory["last_code"].is_type(
-                    "binary_operator", "comparison_operator"
-                ):
-                    # It's not an operator, but the last code was.
-                    if self.operator_new_lines != "before":  # type: ignore
-                        # If in "after" mode, then check to see
-                        # there is a newline between us and the last operator.
-                        for s in memory["since_code"]:
-                            if s.name == "newline":
-                                # Had a newline - so mark last operator as a fail
-                                anchor = memory["last_code"]
-                                # TODO: Work out a nice fix for this failure.
-                # Prepare memory for later
-                memory["last_code"] = context.segment
-                memory["since_code"] = []
-            else:
-                # This isn't a code segment...
-                # Prepare memory for later
-                memory["since_code"].append(context.segment)
-        else:
-            # Reset the memory if we're not in an expression
-            memory = {"last_code": None, "since_code": []}
-
-        # Anchor is our signal as to whether there's a problem
-        if anchor:
-            return LintResult(anchor=anchor, memory=memory, description=description)
-        else:
-            return LintResult(memory=memory)
+        relevent_types = ["binary_operator", "comparison_operator"]
+        segment = context.functional.segment
+        # bring var to this scope so as to only have one type ignore
+        operator_new_lines: str = self.operator_new_lines  # type: ignore
+        expr = segment.children()
+        operator_segments = segment.children(sp.is_type(*relevent_types))
+        results: List[LintResult] = []
+        # If len(operator_segments) == 0 this will essentially not run
+        for operator in operator_segments:
+            start = expr.reversed().first(FirstCodeAfterOperator(operator))
+            end = expr.first(FirstCodeAfterOperator(operator))
+            res = [
+                expr.select(start_seg=start.get(), stop_seg=operator),
+                expr.select(start_seg=operator, stop_seg=end.get()),
+            ]
+            # anchor and change els are reversed in the before case
+            if operator_new_lines == "before":
+                res = [els.reversed() for els in reversed(res)]
+
+            change_list, anchor_list = res
+            # If the anchor side of the list has no newline
+            # then everything is ok already
+            if not anchor_list.any(sp.is_name("newline")):
+                continue
+
+            insert_anchor = anchor_list.last().get()
+            assert insert_anchor, "Insert Anchor must be present"
```

Comment:
What could cause this to fail? I'm wondering if we should handle it without raising an exception. (I'm remembering past bugs that arose from "surprises" in someone's SQL we were processing.)

Reply:
Im using an assert cause it cant happen. This is just for the typechecking / gaurding.
Assert prevents making a dead branch and thus reducing the coverage

Reply:
call signature of .get() has a None, but it will never happen here. If it did it would probably only be regression and not a runtime thing.

Reply:
👍